### ISSUE TITLE
Improve error reporting: Use before definition within files and across whole project

### ISSUE THREAD

**Issue by dsyme** [dotnet/fsharp#12341]
We should give good and clear diagnostics on "use before definition" both within a single file and across the whole project.

The F# compiler parses all files in the project before checking anything.  The compiler service can also be adjusted to parse all files

Given that, it's easy to extract all potential top-level symbols across all files (and the language service has some of those)

If an undefined symbol appears in this set, we can thus give much better diagnsotics, explaining to the user that they need to define functions before they use them.

This is perhaps the most important diagnostic we need to implement, as research by @cartermp showed that this problem was the most common confusion and rejection point for those learning F#.

**Comment by smoothdeveloper** [comment]
This is going to be helpful to people learning the language, but having not experienced environments where it is necessary to have "forward declarations" (.h in C/C++ for example) or just ordering of the code like in F#.

When you are familiar with vanilla OO languages, and not grasped/embraced yet what it does to the code to have this default (no cyclic dependency for vast majority of it), it is a sticking point, and the error message pointing to the "why"s will smoothen embracing this aspect of the language.

**Comment by dsyme** [comment]
Yes, I'm sure the error message will need to include a link to a https://docs.microsoft.com/dotnet/fsharp article explaining the rationales

**Comment by cartermp** [comment]
Completely agree. This was the main thing people new to F# (including without C# experience) ran into.

**Comment by cartermp** [comment]
And to add, once they understood what was going on, it was not a problem for them. So a clear diagnostic is a huge win.

**Comment by smoothdeveloper** [comment]
> as research by @cartermp showed that this problem was the most common confusion and rejection point for those learning F#.

@cartermp, you've aced on all this usability/research work, more people will squint their eyes, witnessing robust & concise F# code, and adopt writing in it in the future, due to compiler & tooling friendliness that keeps piling :)

**Comment by auduchinok** [comment]
We could also try to resolving names using different resolution rules as a last chance, providing better errors for, say, module name used where an expression is expected. That would much improve experience while editing code and looking at unfinished expressions or patterns.

```fsharp
module Nested =
    let x = 1

Nested
```

<img width="363" alt="Screenshot 2021-11-19 at 15 35 13" src="https://user-images.githubusercontent.com/3923587/142627770-872e7c95-6e6b-4bc2-ace5-85822d39526b.png">
<img width="365" alt="Screenshot 2021-11-19 at 15 35 19" src="https://user-images.githubusercontent.com/3923587/142627780-f47d2f0b-14f8-4320-b668-48e64c3019f6.png">
